### PR TITLE
Mc testing to dev

### PR DESCRIPTION
**Pull request by maxachis**
#### Fixes

* Partially addresses  [v2 Testing Overhaul](https://github.com/Police-Data-Accessibility-Project/data-sources-app/issues/271) -- some tests have not been converted to the new setup, and the `docs` repository has not yet been updated to account for the presence of the development database. These components would need to be addressed in a new issue. 

#### Description

* Created new `tests` directory.
* Overhauled the majority of tests formerly in `app_test.py` to test directly on the server development database rather than on a local sqlite database. 
* Refined and expanded most existing tests in `app_test` to account for this development, including adding type hinting and docstrings.
* Moved said tests into separate files (separated by originating script) in a separate `tests/middleware` directory.
* Remaining tests in `app_test`, and their associated test data, have been moved to the `tests/resources` directory -- this should be addressed in a separate issue
* `regular_api_checks`, despite also being tests, has been left untouched -- this should be addressed in a separate issues
* Created a new github action, `test_api.yml`, to perform these tests on every pull request.

#### Testing

* Prior to testing, the `DEV_DB_CONN_STRING`  Github Secret should be added to the associated Github Secrets.
* Testing should be able to be performed by following the instructions in the "Testing" section. 
* Note that one test -- `test_data_source_by_id_query` -- currently fails. After evaluating the test, this appears to be because the database schema deviates from the schema provided in the originating script. In the interests of not making an already considerable PR even larger, I've left this aside for now.

#### Performance

* Longer than previous tests, but still marginal. Currently, on my personal machine, the full suite of 19 tests takes 5 seconds. 

#### Docs

* None

### CODE REVIEW

**Comment by mbodeantor** [comment]
@maxachis The use of the dev database here is confusing to me. Seems like maybe we are missing a fork of the dev repo for V2. Ideally only the dev branches would be pointed at the dev database.

Also can we maintain the use of the DO_DATABASE_URL for all databases? Should make it easier to merge code from dev to main, we can just set the this variable to point at the dev DB in the dev repos and in the dev DO app

**Comment by maxachis** [comment]
> @maxachis The use of the dev database here is confusing to me. Seems like maybe we are missing a fork of the dev repo for V2. Ideally only the dev branches would be pointed at the dev database.

@mbodeantor A little confused by what you mean. This is a PR into the dev repo, which is a fork of the main app, is it not? 

As far as pointing the dev branches at the dev database, a wrinkle arrives when it comes to testing. Even when all this is pushed to production, presumably we still want to test the functionality of the code, but not risk those tests interfering with the production database. If production code fails when testing against the dev database (itself a daily mirror of the production database), then that would clue us in to a problem. So in addition to testing only on PRs, we probably want to run the tests regularly on the database itself.

> Also can we maintain the use of the DO_DATABASE_URL for all databases? Should make it easier to merge code from dev to main, we can just set the this variable to point at the dev DB in the dev repos and in the dev DO app

This is related to my above point -- from my conception, `DEV_DB_CONN_STRING` should _always_ point to the Dev database, because it will always be used for testing. `DO_DATABASE_URL` can point to the dev database for the purposes of testing the full web app, and switch to pointing to prod in the prod environment. But using the prod database for testing (which involves creating and removing data regularly) seems like it runs the risk of interfering with prod data.

**Comment by mbodeantor** [comment]
@maxachis To me, data-sources-app-v2 would be prod, like data-sources-app and we would have dev-v2 or something like that to mirror our current dev, which would be used for testing but fully isolated from our current app structure.

Currently, when we open a PR to prod, the Github action runs tests as a last check to ensure nothing will break the app before merging. All those tests running are already merged into prod and therefore should have little chance of damaging the prod DB. As a dry run for merging, IMO we want to ensure all code runs against the prod DB before going live, even if the dev DB should be an exact copy.

**Comment by maxachis** [comment]
> @maxachis To me, data-sources-app-v2 would be prod, like data-sources-app and we would have dev-v2 or something like that to mirror our current dev, which would be used for testing but fully isolated from our current app structure.
> 
> Currently, when we open a PR to prod, the Github action runs tests as a last check to ensure nothing will break the app before merging. All those tests running are already merged into prod and therefore should have little chance of damaging the prod DB. As a dry run for merging, IMO we want to ensure all code runs against the prod DB before going live, even if the dev DB should be an exact copy.

@mbodeantor The problem with running tests against the production database is that not only many of the endpoints, but many of the tests will interface with the database, which poses a few problems. 

Many of these tests insert fake data. This fake data is designed to be rolled back, but that may not always be successful. This isn’t a problem in the Dev database, because that data is refreshed regularly. But doing such actions in production risks polluting the database with fake data. In fact, we already have this problem – there’s a not-insubstantial number of entries in the database clearly labeled “test”. 

If a test deletes or updates data, the problem becomes more pronounced. Someone could make a test that deletes everything as part of its setup. If we run that in production, well, it’ll delete everything! And we can’t just constrain the role of the connection string so it doesn’t delete or update data, because some of our endpoints delete or update data.  

I’m all for sanity testing production, but we probably shouldn’t have sanity tests include inserts, updates, or deletes. If we’re worried about the test environment not properly emulating production, then we should design the test environment to properly emulate production, not test on production. If we’re running tests to make sure something doesn’t break the app, we probably don’t want the outcome of the test to itself break the app.

**Comment by maxachis** [comment]
@mbodeantor After doing a bit more thinking and research on the subject, I got a sense of the general industry consensus on production vs. development testing. 

Generally speaking, most testing occurs in the pre-production environment, which should emulate production as closely as possible. Production tests are generally designed to be less disruptive and involve less system manipulation. This means more monitoring, system health checks, and "smoke testing", which are run on a regular basis (not simply at initial deployment) to evaluate the continued health of the system. 

There's also sometimes a demarcation between "deployment" and "release" testing, where deployment means it's technically in the production environment but not exposed to users (typically through the use of feature flags), whereas in release it is then exposed to users. Not everyone agrees with that procedure, but it is also an option.

Regardless, the industry standard generally aligns with different testing approaches for development and production, and very much not running the same tests in both environments. That strongly favors maintaining separate connection strings to the different databases.

**Comment by mbodeantor** [comment]
@maxachis I'm in favor of a development database for the majority of testing, just maintaining our automated tests on PR on the prod database.

@josh-chamberlain Do you have strong feelings about this?

**Comment by josh-chamberlain** [comment]
I think we're _mostly_ saying the same things, but it's hard to have an organized conversation. 

- I would prefer not to do update/delete testing on production, and rather make sure our `dev` database is a faithful mirror.
- I think we can use the API to route to the appropriate database for tests vs production; tests could pass a different db URL
- I don't think we should worry about the `deployment` and `release` testing distinction for now.

name | purpose | testing | database
-- | -- | -- | --
app/main | live use | read-only tests for status checks | `prod-v1`
app/dev | "pre-prod" testing | db-affecting tests OK | sqlite (dev-v1)
app-v2/main | live use (secretly published) | read-only status checks | `prod-v2`, mirrored from `prod-v1` periodically*
app-v2/dev | "pre-v2-prod" testing | db-affecting tests OK | `dev-v2`, mirrored from `prod-v1` periodically

*If we have schema changes and migrations, we will need to make sure those run / stack on top of the v1 database over the course of v2 development.

type of testing | timing | purpose | db
-- | -- | -- | --
v1 automated status checks | async from code | "is the app running?" | prod-v1, read-only
v2 automated status checks | async from code | "is v2 running?" | prod-v2
`app/tests` | locally while coding, PR checks | prevent breaking changes | sqlite
`app-v2/tests` | locally while coding, PR checks | prevent breaking changes | `dev-v2` 

↑ I believe the local tests are the same ones which run on PR check, correct?

File: .github/workflows/test_api.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+name: Test API using Pytest
+
+on:
+  pull_request:
+
+jobs:
+  test_api:
+    env:
+      SECRET_KEY: ${{ secrets.SECRET_KEY }}
+      DEV_DB_CONN_STRING: ${{secrets.DEV_DB_CONN_STRING}}
```
Comment (josh-chamberlain) [inline .github/workflows/test_api.yml:10]: i have added this as an org level secret with the connection string for the new DO data_sources_app_v2 user

**Review by josh-chamberlain — APPROVED** [review]

**Pull request merged**